src/calc_overlap.cpp

```diff
@@ -12,52 +13,218 @@ using namespace Rcpp;
 //' c++ implementation for calculating the cell-neighour-overlapping among k nearest cell neighbours of each gene.
 //' @param cc_adj sparse matrix (dgCMatrix), cell-cell adjacency matrix
 //' @param cg_adj sparse matrix (dgCMatrix), cell-gene adjacency matrix
+//' @param threshold numeric value between 0 and 1. The cutoff of cell-nqighour-overlapping of each gene. 
 //' @export
 // [[Rcpp::export]]
-Eigen::SparseMatrix<double> calc_overlap(Eigen::SparseMatrix<int> cc_adj,
-                                         Eigen::SparseMatrix<int> cg_adj) {
-  
+void calc_overlap(Eigen::Map<Eigen::SparseMatrix<double>>& cc_adj,
+                                         Eigen::Map<Eigen::SparseMatrix<double>>& cg_adj,
+                                        double threshold) {
+
   // initialize vector to store triplets
-  typedef Eigen::Triplet<double> Trip;
-  std::vector<Trip> trp;
-  Eigen::SparseMatrix<int> overlap_mat_all = cc_adj * cg_adj;
-  Eigen::SparseMatrix<int> cc_tadj = cc_adj.transpose();
-  
+
+  Eigen::SparseMatrix<double> overlap_mat_all = cc_adj * cg_adj;
+  Eigen::SparseMatrix<double> cc_tadj = cc_adj.transpose();
+
   // calcualte the rowSums of matrix cc_adj which is also the number of neighbourhoods of each cell
   std::vector<double> cell_nn_nums;
+
   for (int i=0; i < cc_tadj.outerSize(); i++){
-    int k = 0;
-    
-    for (Eigen::SparseMatrix<int>::InnerIterator it(cc_tadj, i); it; ++it){  // Iterate over rows
-      k += 1;
+
+      int k = 0;
+
+      for (Eigen::SparseMatrix<double>::InnerIterator it(cc_tadj, i); it; ++it){  // Iterate over rows
+          k += 1;
     }
-    
-    cell_nn_nums.push_back(k);
+
+      cell_nn_nums.push_back(k);
   }
-  
+
+
+
   // loop over genes (its faster in column major matrix)
   for (int i=0; i < cg_adj.outerSize(); i++){
-    
+
     // only preserve the edges which are shown in cg_adj matrix
-    for (Eigen::SparseMatrix<int>::InnerIterator it(cg_adj, i); it; ++it){  // Iterate over rows
-      
-      double value = overlap_mat_all.coeffRef(it.row(), i)/cell_nn_nums[it.row()];
-      
-      trp.push_back(Trip(it.row(),
-                         i, 
-                         value));
-    }
+    for (Eigen::Map<Eigen::SparseMatrix<double>>::InnerIterator it(cg_adj, i); it; ++it){  // Iterate over rows
+
+        double value = overlap_mat_all.coeffRef(it.row(), i)/cell_nn_nums[it.row()];
+
+        if (value <= threshold){
+
+            it.valueRef() = 0;
+        }
     }
-  
-  // build Matrix from triplets
-  Eigen::SparseMatrix<double> overlap(cg_adj.rows(), cg_adj.cols());
-  overlap.setFromTriplets(trp.begin(), trp.end());
-  
-  return overlap;
+  }
+
+    return;
```

Comment:
this returns nothing, but in graph.R we assign the output to cgg_nn:
```
cgg_nn <- calc_overlap( cc_adj = ccg_nn,
                                     cg_adj = cgg_nn,
                                     threshold = overlap)
```
I guess we should not assign the output to cgg_nn then?